### PR TITLE
{bio}[foss/2019b,fosscuda/2019b] R-bundle-Bioconductor v3.10

### PR DESCRIPTION
**Pull request by edmondac**
(created using `eb --new-pr`)

Requires #9519 to be merged first (done)

Requires easybuilders/easybuild-easyblocks#1938

### CODE REVIEW

**Comment by edmondac** [comment]
Actually, the major problem is that `dada2` uses `cpuid` - and the `cpuid.h` header doesn't exist on various architectures (e.g. ppc64le). It may be possible to patch it, since it's not used very much:
```
./src/Rmain.cpp:5:#define cpuid(info, x)    __cpuidex(info, x, 0)
./src/Rmain.cpp:7:#include <cpuid.h>
./src/Rmain.cpp:8:void cpuid(int info[4], int InfoType){
./src/Rmain.cpp:9:  __cpuid_count(InfoType, 0, info[0], info[1], info[2], info[3]);
./src/Rmain.cpp:84:  cpuid(info, 0);
./src/Rmain.cpp:89:    cpuid(info,0x00000001);
```

But for now I'll just remove `dada2`

**Comment by edmondac** [comment]
Errors from LEA:
```
In file included from sNMF/sNMF.h:13,
                 from R_sNMF.c:24:
sNMF/../bituint/bituint.h:141:2: error: #error Unsupported architecture
```

**Comment by zao** [comment]
There does not seem to be any `MariaDB-connector-c/2.3.7` compatible with these GCC 8.3.0-based toolchains.

**Comment by edmondac** [comment]
Need to add metap here: see https://github.com/easybuilders/easybuild-easyconfigs/pull/9519#issuecomment-573016858

**Comment by edmondac** [comment]
> There does not seem to be any `MariaDB-connector-c/2.3.7` compatible with these GCC 8.3.0-based toolchains.

Yes - sorry, I'd missed "WIP" off the title. I've not added everything yet.

**Comment by boegel** [comment]
@edmondac That 20k-line patch file for `lpsymphony` is pretty nasty...

File: easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-foss-2019b-R-3.6.2.eb [hunk 1]
```diff
@@ -0,0 +1,1430 @@
...
+# (some versions in this bundle may be newer than the ones provided by R)
+local_ext_version_check = "pkgver = packageVersion('%(ext_name)s'); if (pkgver != '%(ext_version)s') "
+local_stop_msg = "stop('%(ext_name)s %(ext_version)s not installed, found ', pkgver, ' instead')"
+exts_filter = ("R -q --no-save", "%s { %s }" % (local_ext_version_check, local_stop_msg))
+
+# CRAN packages on which these Bioconductor packages depend are available in R module on which this depends
+# !! order of packages is important !!
+# packages updated on Aug 13th 2019
+exts_list = [
+    ('Rcpp', '1.0.3', {
```
Comment (boegel) [inline easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-foss-2019b-R-3.6.2.eb]: @edmondac Some of these extensions are already provided by the `R` installation, they don't need to be included again here?

Examples: `Rcpp`, `BH`, `crayon`, `rlang`, ...

**Comment by edmondac** [comment]
> @edmondac That 20k-line patch file for `lpsymphony` is pretty nasty...

I was wondering if there could be a nicer pattern for this. They have 7 or 8 copies of config.guess within the extension tgz, and I wondered if there was a mechanism for copying the standard easybuild config.guess over the top without doing a patch ...?

File: easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-foss-2019b-R-3.6.2.eb (hunk 1 above)
Comment (edmondac) [reply easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-foss-2019b-R-3.6.2.eb]: That'll be easy_update - I'll weed out the duplicates

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 8 out of 8 (5 easyconfigs in this PR)
bear-pg0212u17a.bear.cluster - Linux centos linux 7.7.1908, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 2.7.5
See https://gist.github.com/95ff58244e160ea4d08e280abaef6de5 for a full test report.

**Comment by boegel** [comment]
> > @edmondac That 20k-line patch file for `lpsymphony` is pretty nasty...
> 
> I was wondering if there could be a nicer pattern for this. They have 7 or 8 copies of config.guess within the extension tgz, and I wondered if there was a mechanism for copying the standard easybuild config.guess over the top without doing a patch ...?

We could add some logic to the `RPackage` easyblock to pull in (on-demand) a recent `config.guess` and use that rather the one that was shipped (like we do in `ConfigureMake`).

Or you could contribute an updated `config.guess` to upstream, and ask for a new release on CRAN? Long term, that's certainly the best option...

File: easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-foss-2019b-R-3.6.2.eb [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+
+# check whether correct version is installed in extension filter
+# (some versions in this bundle may be newer than the ones provided by R)
+local_ext_version_check = "pkgver = packageVersion('%(ext_name)s'); if (pkgver != '%(ext_version)s') "
+local_stop_msg = "stop('%(ext_name)s %(ext_version)s not installed, found ', pkgver, ' instead')"
+exts_filter = ("R -q --no-save", "%s { %s }" % (local_ext_version_check, local_stop_msg))
+
+# CRAN packages on which these Bioconductor packages depend are available in R module on which this depends
+# !! order of packages is important !!
+# packages updated on Aug 13th 2019
```
Comment (boegel) [inline easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-foss-2019b-R-3.6.2.eb]: date needs a bump?

File: easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-fosscuda-2019b-R-3.6.2.eb [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+
+# check whether correct version is installed in extension filter
+# (some versions in this bundle may be newer than the ones provided by R)
+local_ext_version_check = "pkgver = packageVersion('%(ext_name)s'); if (pkgver != '%(ext_version)s') "
+local_stop_msg = "stop('%(ext_name)s %(ext_version)s not installed, found ', pkgver, ' instead')"
+exts_filter = ("R -q --no-save", "%s { %s }" % (local_ext_version_check, local_stop_msg))
+
+# CRAN packages on which these Bioconductor packages depend are available in R module on which this depends
+# !! order of packages is important !!
+# packages updated on Aug 13th 2019
```
Comment (boegel) [inline easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-fosscuda-2019b-R-3.6.2.eb]: date needs a bump

**Comment by edmondac** [comment]
> We could add some logic to the `RPackage` easyblock to pull in (on-demand) a recent `config.guess` and use that rather the one that was shipped (like we do in `ConfigureMake`).

I'll have a look at that and see if it's easy to do...

File: easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-foss-2019b-R-3.6.2.eb (hunk 1 above)
Comment (edmondac) [reply easybuild/easyconfigs/r/R-bundle-Bioconductor/R-bundle-Bioconductor-3.10-foss-2019b-R-3.6.2.eb]: Done

**Comment by edmondac** [comment]
Possible solution @boegel :

1. Extract functions `obtain_config_guess`, `check_config_guess` and the `CONFIG_GUESS_*` constants into a new file, perhaps `_config_guess.py` in easyblocks/generic - or should it be in the framework?
2. Make configuremake.py use the new file
3. Make rpackage.py use the new file

It doesn't look like this would take too long to do... what do you think?

**Comment by edmondac** [comment]
I've made a prototype implementation and am testing it...

**Comment by edmondac** [comment]
https://github.com/easybuilders/easybuild-easyblocks/pull/1937 is the start of a PR for the config.guess replacement code - but I need to clean it up.

**Comment by verdurin** [comment]
@edmondac Do you consider this ready for testing once the easyblock PR is merged?

**Comment by edmondac** [comment]
> @edmondac Do you consider this ready for testing once the easyblock PR is merged?

Yes - it's good to go, I think. Once https://github.com/easybuilders/easybuild-easyblocks/pull/1938 is merged I can upload some test reports.

**Comment by boegel** [comment]
@edmondac This needs a sync with current `develop` now #9941 is merged...

**Comment by edmondac** [comment]
This PR has sat around too long and the world's moved on past it.

This had some POWER9 fixes in it, but this probably needs to be re-done against the already  merged R-bundle-Bioconductor-3.10-foss-2019b.eb :-(

**Comment by verdurin** [comment]
@edmondac How would you like to proceed here, given the developments elsewhere that you've noted?

**Pull request closed without merging**